pkg/kubevirt/client.go

```diff
@@ -180,3 +224,155 @@ func (c *client) DeleteDataVolume(ctx context.Context, namespace string, name st
 	return c.cdiClient.CdiV1beta1().DataVolumes(namespace).Get(ctx, name, metav1.GetOptions{})
 }
+
+func (c *client) CreateVolumeSnapshot(ctx context.Context, namespace, name, claimName, snapshotClassName string) (*snapshotv1.VolumeSnapshot, error) {
+	if dv, err := c.GetDataVolume(ctx, namespace, claimName); err != nil {
+		return nil, err
+	} else if dv != nil {
+		snapshotClassName, err := c.getSnapshotClassNameFromVolumeClaimName(ctx, namespace, dv.GetName(), snapshotClassName)
+		if err != nil {
+			return nil, err
+		}
+		snapshot := &snapshotv1.VolumeSnapshot{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: namespace,
+				Labels:    c.infraLabelMap,
+			},
+			Spec: snapshotv1.VolumeSnapshotSpec{
+				Source: snapshotv1.VolumeSnapshotSource{
+					PersistentVolumeClaimName: &claimName,
+				},
+			},
+		}
+		if snapshotClassName != "" {
+			snapshot.Spec.VolumeSnapshotClassName = &snapshotClassName
+		} else {
+			return nil, fmt.Errorf("no snapshot class found for infra source volume")
+		}
+		klog.V(5).Infof("Creating snapshot %s with snapshot class %s, %#v", name, snapshotClassName, snapshot)
+		return c.snapClient.SnapshotV1().VolumeSnapshots(namespace).Create(ctx, snapshot, metav1.CreateOptions{})
+	}
+	return nil, nil
+}
+
+func (c *client) getSnapshotClassNameFromVolumeClaimName(ctx context.Context, namespace, claimName, snapshotClassName string) (string, error) {
+	volumeName, err := c.getVolumeNameFromClaimName(ctx, namespace, claimName)
+	if err != nil || volumeName == "" {
+		klog.V(2).Infof("Error getting volume name for claim %s in namespace %s: %v", claimName, namespace, err)
+		return "", fmt.Errorf("unable to determine snapshot class name for infra source volume")
+	}
+	storageClassName, err := c.getStorageClassFromVolume(ctx, volumeName)
+	if err != nil {
+		return "", err
+	}
+	snapshotClass, err := c.getSnapshotClassFromStorageClass(ctx, storageClassName, snapshotClassName)
+	if err != nil {
+		return "", err
+	}
+	return snapshotClass.Name, nil
+}
+
+// Determine the name of the volume associated with the passed in claim name
+func (c *client) getVolumeNameFromClaimName(ctx context.Context, namespace, claimName string) (string, error) {
+	volumeClaim, err := c.kubernetesClient.CoreV1().PersistentVolumeClaims(namespace).Get(ctx, claimName, metav1.GetOptions{})
+	if err != nil {
+		klog.Errorf("Error getting volume claim %s in namespace %s: %v", claimName, namespace, err)
+		return "", err
+	}
+	klog.V(5).Infof("found volumeClaim %#v", volumeClaim)
+	return volumeClaim.Spec.VolumeName, nil
+}
+
+// Determine the storage class from the volume
+func (c *client) getStorageClassFromVolume(ctx context.Context, volumeName string) (string, error) {
+	volume, err := c.kubernetesClient.CoreV1().PersistentVolumes().Get(ctx, volumeName, metav1.GetOptions{})
+	if err != nil {
+		klog.V(2).Infof("Error getting volume %s: %v", volumeName, err)
+		return "", err
+	}
+	return volume.Spec.StorageClassName, nil
+}
```

Comment:
I think it would be wise for us to perform a check on the volume's storageclass to make sure it is within our allow list.

example: https://github.com/kubevirt/csi-driver/blob/7aa580f8933c1ffb54f12aa5985781fa7e84f308/pkg/service/controller.go#L83

Reply:
Yes good point, adding check